src/Guard/TelegramAuthenticator.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+namespace BoShurik\TelegramBotBundle\Guard;
+
+use BoShurik\TelegramBotBundle\Guard\UserProviderInterface as BoShurikUserProviderInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\Authenticator\AbstractFormLoginAuthenticator;
+use Symfony\Component\Security\Http\Util\TargetPathTrait;
+
+class TelegramAuthenticator extends AbstractFormLoginAuthenticator
+{
+    use TargetPathTrait;
+
+    private const REQUIRED_FIELDS = [
+        'id',
+        'first_name',
+        'last_name',
+        'auth_date',
+        'hash',
+    ];
+
+    /**
+     * @var string
+     */
+    private $secret;
+
+    /**
+     * @var BoShurikUserProviderInterface
+     */
+    private $userProvider;
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    public function __construct(BoShurikUserProviderInterface $userProvider, UrlGeneratorInterface $urlGenerator, string $telegramBotToken)
+    {
+        $this->userProvider = $userProvider;
+        $this->urlGenerator = $urlGenerator;
+        $this->secret = hash('sha256', $telegramBotToken, true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request)
+    {
+        $route = $request->attributes->get('_route');
+        $data = $request->query->all();
+
+        return $route === '_telegram_login'
+            && array_intersect(self::REQUIRED_FIELDS, array_keys($data)) === self::REQUIRED_FIELDS;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    final public function getCredentials(Request $request)
+    {
+        return $request->query->all();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUser($credentials, UserProviderInterface $userProvider)
+    {
+        $this->validate($credentials);
+
+        return $this->userProvider->loadUserByTelegramId($credentials);
+    }
+
+    /**
+     * Check the data integrity.
+     */
+    final protected function validate(array $data)
```

Comment:
I suggest to move it in separate `DataValidator` class, so it can be reused in cases when we want to attach `telegramId ` to already logged in user

Reply:
Done! Exported into `TelegramLoginValidator` class.